src/main/java/org/jenkinsci/plugins/workflow/steps/TimeoutStepExecution.java

```diff
@@ -256,6 +263,34 @@ private static ExecutorService newExecutorService() {
             }
         }
 
+        @Override
+        public void onFailure(StepContext context, Throwable t) {
+            super.onFailure(context, t);
+            if (t instanceof FlowInterruptedException) {
+                /*
+                 * Check whether the FlowInterruptedException has propagated past the corresponding
+                 * timeout step.
+                 */
+                FlowNode flowNode = null;
+                try {
+                    flowNode = getContext().get(FlowNode.class);
+                } catch (IOException | InterruptedException e) {
+                    LOGGER.log(Level.WARNING, null, e);
+                }
+                if (flowNode != null) {
+                    String nodeId = flowNode.getId();
+                    for (CauseOfInterruption cause : ((FlowInterruptedException) t).getCauses()) {
+                        if (cause instanceof ExceededTimeout) {
+                            ExceededTimeout exceededTimeout = (ExceededTimeout) cause;
+                            if (nodeId.equals(exceededTimeout.getNodeId())) {
```

Comment:
nit: Maybe worth updating the comment on `ExceededTimeout.nodeId` to mention that `exceededTimeout.getNodeId()` could be `null` for exceptions serialized before this change. Everything should work fine though since you're not using it as the receiver here, so it would just match the behavior from before this change.

Reply:
Done